### PR TITLE
android-ndk hook fixes + test_package

### PR DESCRIPTION
**Pull request by madebr**
Let's get this working.
https://github.com/conan-io/conan-center-index/issues/3996 was opened today because bincrafters' android-ndk-installer recipe is outdated.
It looks like https://github.com/conan-io/conan-center-index/pull/3004 is stalled.

- this pr fixes the hook warnings/errors (on my system).
There is still a FIXME, but let's discuss that on https://github.com/conan-io/conan-center-index/pull/3004
- add a little `ndk-build` test_package. Too bad there is no way to add a cmake test_package (that I know of).
This is really missing from conan.

I tested this as `conan create . android-ndk/r21d@` + building `zlib/1.2.11` in cross building mode, only on Linux.
Not tested on Macos or Windows.

### CODE REVIEW

File: recipes/android-ndk/all/test_package/jni/test_package.c [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#include <android/log.h>
+
+__attribute__((visibility("default")))
+void test_package_log_message(const char *msg) {
+    __android_log_print(ANDROID_LOG_DEBUG, "test_pacakge", "%s", msg);
```
Comment (madebr) [inline recipes/android-ndk/all/test_package/jni/test_package.c:5]:
```suggestion
    __android_log_print(ANDROID_LOG_DEBUG, "test_package", "%s", msg);
```

**Comment by a4z** [comment]
https://github.com/conan-io/conan-center-index/pull/3004 is not stalled,  it just takes time

there are quite many changes, for example ' into " , why ?  this makes it super hard to follow what actually changed.

It will take me quite a while to go through all the changes, what I have currently works for building a lot of packages, rebuild everything and test with those changes takes time

File: recipes/android-ndk/all/conanfile.py [hunk 2]
```diff
@@ -15,198 +15,222 @@ class AndroidNDKInstallerConan(ConanFile):
...
+        bin_path = os.path.join(self._packaged_tar_root, "bin")
+        self.output.info("Appending PATH environment variable: {}".format(bin_path))
+        self.env_info.PATH.append(bin_path)
+
+        self.output.info("Creating ANDROID_NDK_HOME environment variable: %s" % self.package_folder)
+        self.env_info.ANDROID_NDK_HOME = self._packaged_tar_root
+
+        # FIXME: should `android-api` be added to exceptions of KB-H020?
+        self.cpp_info.builddirs.extend([
+            os.path.join("bin", "build", "cmake"),
```
Comment (a4z) [inline recipes/android-ndk/all/conanfile.py:176]: the files in there are added from the toolchain file, so I am not sure it is required adding that into the path, and therefore it seems to me like a non required stepp 

no I can not control all the other if they are required, can you please add comments why the should be in path ?

File: recipes/android-ndk/all/conanfile.py [hunk 3]
```diff
@@ -15,198 +15,222 @@ class AndroidNDKInstallerConan(ConanFile):
...
+            if self.settings_target.arch not in available_target_arch:
+                raise ConanInvalidConfiguration("android-ndk only supports the following target architectures: {}".format(available_target_arch))
+            if self.settings_target.compiler != "clang":
+                raise ConanInvalidConfiguration("android-ndk only supports the clang compiler")
+            if self.settings_target.os != "Android":
+                raise ConanInvalidConfiguration("android-ndk only supports target os=Android")
 
-    def source(self):
+    def build(self):
+        # Since we're downloading settings dependent binaries, download in `build` instead of `source`
```
Comment (a4z) [inline recipes/android-ndk/all/conanfile.py:44]: is this tested that it works with the DL cache, because the nsdk is quite huge ....

File: recipes/android-ndk/all/conanfile.py [hunk 4]
```diff
@@ -15,198 +15,222 @@ class AndroidNDKInstallerConan(ConanFile):
     no_copy_source = True
     exports_sources = ["cmake-wrapper.cmd", "cmake-wrapper"]
 
-    settings = {"os": ["Windows", "Linux", "Macos"],
-                "arch": ["x86_64"]}
+    settings = "os", "arch"
```
Comment (a4z) [inline recipes/android-ndk/all/conanfile.py:18]: why replacing what comes for free from conan, checking the values, with own code in configure ? 
what is the advantage?

**Review by a4z — COMMENTED** [review]
Thanks for you help, @madebr ! 

I have a few questions, and than I need to test that changes will all the recipes I already built, now over Christmas and new year, that might take some days.

And I also did not see any mail notification that this PR has been opened, so I saw it just today, sorry for the delay !

File: recipes/android-ndk/all/conanfile.py [hunk 5]
```diff
@@ -15,198 +15,222 @@ class AndroidNDKInstallerConan(ConanFile):
...
                          sig == [0xFE, 0xED, 0xFA, 0xCF] or \
                          sig == [0xCF, 0xFA, 0xED, 0xFE] or \
                          sig == [0xFE, 0xEF, 0xFA, 0xCE] or \
                          sig == [0xCE, 0xFA, 0xED, 0xFE]:
-                        self.output.info('chmod on Mach-O file: "%s"' % filename)
+                        self.output.info("chmod on Mach-O file: '%s'" % filename)
                         self._chmod_plus_x(filename)
 
+    @property
+    def _packaged_tar_root(self):
```
[the remark this answers is not in the document]
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:127]: This is to circumvent the hooks only allowing `bin`, `lib`, `licenses` and `res` in the root folder.

File: recipes/android-ndk/all/conanfile.py (hunk 2 above)
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:176]: When not doing this, KB-H020 will complain about `.cmake` files.
I *think* adding them to `cpp_info.builddirs` will add the folders to `CMAKE_MODULE_PATH` (not to the `PATH` env var).

That's why I added the FIXME on top to generate some discussion.
At the same times, adding the folders to `cpp_info.builddirs` made it pass the hook tests.

File: recipes/android-ndk/all/conanfile.py [hunk 6]
```diff
@@ -15,198 +15,222 @@ class AndroidNDKInstallerConan(ConanFile):
...
+        # FIXME: should `android-api` be added to exceptions of KB-H020?
+        self.cpp_info.builddirs.extend([
+            os.path.join("bin", "build", "cmake"),
+            os.path.join("bin", "sources", "cxx-stl", "llvm-libc++abi"),
+            os.path.join("bin", "bin", "sources", "third_party", "googletest", "cmake", "internal_utils.cmake"),
+            os.path.join("bin", "sources", "third_party", "googletest", "cmake", "internal_utils.cmake"),
+        ])
+
+        # The settings hereafter depend on the target system.
+        if hasattr(self, "settings_target") and self.settings_target is not None:
```
[the remark this answers is not in the document]
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:183]: I guess it's a difference approach.

I avoid a return halfway in a function to avoid people adding things at the end, and them thinking that it will run.

File: recipes/android-ndk/all/conanfile.py [hunk 7]
```diff
@@ -15,198 +15,222 @@ class AndroidNDKInstallerConan(ConanFile):
...
+            self.output.info("Creating NDK_ROOT environment variable: %s" % self._ndk_root)
```
[the remark this answers is not in the document]
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:184]: Is there a difference with the bincrafters recipe?
https://github.com/bincrafters/conan-android_ndk_installer/blob/43ad339549487fbce7cb07ca99b52c32c59559e5/conanfile.py#L157-L158

I did a grep for `NDK_ROOT` in the package folder and got some matches.
I think it is used by ndk-build (which is a collection of make scripts).

File: recipes/android-ndk/all/test_package/conanfile.py [hunk 8]
```diff
@@ -1,12 +1,27 @@
 from conans import ConanFile, tools
 import os
+import shutil
 
 
 class TestPackgeConan(ConanFile):
     settings = "os", "arch"
 
+
```
[the remark this answers is not in the document]
Comment (madebr) [reply recipes/android-ndk/all/test_package/conanfile.py:9]: Indeed, it only tests the simple `conan create . android-ndk/r21d@`.
It would have been nice to test the cmake integration...

But, better then nothing?

File: recipes/android-ndk/all/conanfile.py (hunk 3 above)
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:44]: It caches, at least when you configured `storage.download_cache` in your conan config.

File: recipes/android-ndk/all/conanfile.py (hunk 4 above)
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:18]: Better error messages:
See https://github.com/conan-io/conan-center-index/pull/841#issuecomment-585160197
and https://github.com/conan-io/conan-center-index/pull/842

Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:18]: Also, https://github.com/conan-io/conan/issues/6534

File: recipes/android-ndk/all/conanfile.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (a4z) [reply recipes/android-ndk/all/conanfile.py:184]: seems I looked wrong , the 
```
        self.output.info('Creating ANDROID_NDK_HOME environment variable: %s' % self.package_folder)
        self.env_info.ANDROID_NDK_HOME = self.package_folder
```
seems to be missing, I guess that confused me.

is omitting that intentional?

File: recipes/android-ndk/all/conanfile.py (hunk 4 above)
Comment (a4z) [reply recipes/android-ndk/all/conanfile.py:18]: Ok, but , why not waiting until conan fixes their error messages ? 
and until then ... it is like it is 
This recipe is already somehow complex, each line of non existing code is an improvement, imho, especially if it is not absolute required

File: recipes/android-ndk/all/conanfile.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:184]: It's set before the
```
if hasattr(self, "settings_target") and self.settings_target is not None:
```

`NDK_ROOT` is dependent on the target os/arch, `ANDROID_NDK_HOME` isn't.

File: recipes/android-ndk/all/conanfile.py (hunk 4 above)
Comment (madebr) [reply recipes/android-ndk/all/conanfile.py:18]: I guess it's again a matter of opinion.
I kind of like to group all checks in one location, which here is the `configure` method.

File: recipes/android-ndk/all/conanfile.py [hunk 9]
```diff
@@ -15,198 +15,222 @@ class AndroidNDKInstallerConan(ConanFile):
...
+            if self.settings_target.os != "Android":
+                raise ConanInvalidConfiguration("android-ndk only supports target os=Android")
 
-    def source(self):
+    def build(self):
+        # Since we're downloading settings dependent binaries, download in `build` instead of `source`
         tarballs = self.conan_data["sources"][self.version]["url"]
         tools.get(**tarballs[str(self.settings.os)])
         extracted_dir = self.name + "-" + self.version
         os.rename(extracted_dir, self._source_subfolder)
```
Comment (madebr) [inline recipes/android-ndk/all/conanfile.py:48]: We don't do any patching in the `build` folder and only copy the files from the build to package folder in `package`.
Doing the download + extract in `package` would avoid the copy of 4+GiB on Linux.

Comment (a4z) [reply recipes/android-ndk/all/conanfile.py:48]: yes, I resolved the conversation because I plan to take that change, sorry for not being clear and let you know

**Comment by madebr** [comment]
After https://github.com/conan-io/hooks/pull/263, there are no more hook warnings.
